File: python/leetcode/1614.py
from typing import List


class Solution:
    def maxDepth(self, s: str) -> int:
        stack: List[int] = []
        max_depth: int = 0
        depth_count: int = 0

        for ch in s:
            if ch == "(":
                depth_count += 1
                max_depth = max(depth_count, max_depth)
                stack.append(ch)
            elif ch == ")":
                if stack[-1] == "(":
                    depth_count -= 1
                    stack.pop()

        return max_depth


if __name__ == "__main__":
    # s = "(1+(2*3)+((8)/4))+1"
    s = "(1)+((2))+(((3)))"
    solution = Solution()
    print(solution.maxDepth(s))

File: cpp/leetcode/1614.cpp
#include <iostream>
#include <stack>
#include <algorithm>

class Solution {
public:
    int maxDepth(std::string s) {
        std::stack<char> chStack;

        int depthCount = 0;
        int maxCount = 0;

        for (auto& ch : s) {
            if (ch == '(') {
                chStack.push(ch);
                depthCount += 1;
                maxCount = std::max(maxCount, depthCount);
            } else if (ch == ')') {
                depthCount -= 1;
                chStack.pop();
            }
        }

        return maxCount;
    }
};

int main() {
    std::string s = "(1)+((2))+(((3)))";
    Solution solution;
    std::cout << solution.maxDepth(s) << std::endl;

    return 0;
}
